src/node.rs

```diff
@@ -0,0 +1,171 @@
+    pub async fn socket(&self) -> io::Result<UdpSocket> {
+        let table_record = self.table.record;
+        let socket_addr = SocketAddrV4::new(table_record.ip_address, table_record.udp_port);
+        let socket = UdpSocket::bind(socket_addr).await;
+        socket
+    }
+
+    // TODO: start_server() -> Result<(), Box<dyn std::error::Error>> {
+    pub async fn start_server(&self, mut buffer: [u8; 4], socket: UdpSocket) {
+        println!("Starting server");
+        loop {
+            let Ok((size, addr)) = socket.recv_from(&mut buffer).await else { todo!() };
+            let message = buffer.clone().to_vec();
+            self.process(message).await;
+        }
+    }
+
+    async fn process(&self, message: Vec<u8>) {
+        println!("Message: {:?}", message);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::LockResult;
+
+    use crate::{helper::PING_MESSAGE_SIZE, node};
+
+    use super::*;
+
+    fn mk_nodes(n: u8) -> (Node, Vec<Node>) {
+        let ip_address = String::from("127.0.0.1").parse::<Ipv4Addr>().unwrap();
+        let port_start = 9000_u16;
+
+        let local_node = mk_node(&ip_address, port_start, 0);
+        let remote_nodes: Vec<Node> = (1..n)
+            .map(|i| mk_node(&ip_address, port_start, i))
+            .collect();
+
+        (local_node, remote_nodes)
+    }
+
+    fn mk_node(ip_address: &Ipv4Addr, port_start: u16, index: u8) -> Node {
+        let mut node_id = [0_u8; 32];
+        node_id[31] += index;
+        let udp_port = port_start + index as u16;
+
+        let table_record = TableRecord {
+            ip_address: *ip_address,
+            udp_port,
+        };
+
+        Node::new(node_id, table_record)
+    }
+
+    #[test]
+    fn add_redundant_node() {
+        let (mut local_node, remote_nodes) = mk_nodes(2);
+
+        let result = local_node
+            .table
+            .add(remote_nodes[0].table.id, remote_nodes[0].table.record);
+        assert!(result);
+        let result2 = local_node
+            .table
+            .add(remote_nodes[0].table.id, remote_nodes[0].table.record);
+        assert!(!result2);
+    }
+
+    #[test]
+    fn find_node() {
+        let (mut local_node, remote_nodes) = mk_nodes(10);
+        let node_to_find = &remote_nodes[1];
+        let ntf_bucket_index = local_node.table.xor_bucket_index(&node_to_find.table.id);
+
+        for node in &remote_nodes {
+            local_node.table.add(node.table.id, node.table.record);
+        }
+
+        let closest_nodes = local_node.find_node(&node_to_find.node_id);
+        for node in closest_nodes {
+            let bucket_index = local_node.table.xor_bucket_index(&node.0);
+            assert_eq!(ntf_bucket_index, bucket_index);
+        }
+    }
+
+    #[tokio::test]
+    async fn ping() {
+        let (mut local_node, remote_nodes) = mk_nodes(2);
+        local_node
+            .table
+            .add(remote_nodes[0].table.id, remote_nodes[0].table.record);
```

Comment:
you'll prolly want a higher-level interface for this, like `add_peer(peer: Peer)` or however you want to pass in the node's data

```rust
struct Peer {
    id: Id,
    ip: IpType,
    port: u16
}
```